src/main/java/com/google/cloud/genomics/dataflow/functions/JoinGvcfVariants.java

```diff
@@ -0,0 +1,189 @@
+
+    public long getStartBin(int binSize, Variant variant) {
+      // Round down to the nearest integer
+      return Math.round(Math.floor(variant.getStart() / binSize));
+    }
+
+    public long getEndBin(int binSize, Variant variant) {
+      // Round down to the nearest integer
+      return Math.round(Math.floor(variant.getEnd() / binSize));
+    }
+
+    @Override
+    public void processElement(ProcessContext context) {
+      GenomicsDatasetOptions options =
+          context.getPipelineOptions().as(GenomicsDatasetOptions.class);
+      Variant variant = context.element();
+      for (long startBin = getStartBin(options.getBinSize(), variant), endBin =
+          VariantUtils.isVariant(variant) ? startBin : getEndBin(options.getBinSize(), variant), bin =
+          startBin; bin <= endBin; bin++) {
+        context.output(KV.of(KV.of(variant.getReferenceName(), bin), variant));
+      }
+    }
+  }
+
+  @SuppressWarnings("serial")
+  public static class MergeVariants extends DoFn<KV<KV<String, Long>, Iterable<Variant>>, Variant> {
+
+    @Override
+    public void processElement(ProcessContext context) {
+
+      // The upper bound on number of variants in the iterable is dependent upon the binSize
+      // used in the prior step to construct the key.
+      KV<KV<String, Long>, Iterable<Variant>> kv = context.element();
+
+      // Sort by start position descending and ensure that if a variant
+      // and a ref-matching block are at the same position, the
+      // ref-matching block comes first.
+      List<Variant> records = Lists.newArrayList(kv.getValue());
+      Collections.sort(records, new JoinGvcfVariants.VariantComparator());
+
+      // The upper bound on potential overlaps is the sample size plus the number of
+      // block records that occur between actual variants.
+      List<Variant> blockRecords = new LinkedList<Variant>();
+
+      for (Variant record : records) {
+        if (VariantUtils.isVariant(record)) {
+          // TODO: determine and implement the correct criteria for overlaps of non-SNP variants
+          if (VariantUtils.isSnp(record)) {
+            for (Iterator<Variant> iterator = blockRecords.iterator(); iterator.hasNext();) {
+              Variant blockRecord = iterator.next();
+              if (JoinGvcfVariants.isOverlapping(blockRecord, record)) {
+                record.getCalls().addAll(blockRecord.getCalls());
+              } else {
+                // Remove the current element from the iterator and the list since it is
+                // left of the genomic region we are currently working on due to our sort..
+                iterator.remove();
+              }
+            }
+          }
+          context.output(record);
+        } else {
+          blockRecords.add(record);
+        }
+      }
+    }
+  }
+
+  /**
+   * Comparator for sorting variants by genomic position and alternateBases
+   */
+  public static class VariantComparator implements Comparator<Variant> {
```

Comment:
A guava Ordering would be much cleaner than this handrolled Comparator.

If you aren't going to use Ordering, then at minimum i'd suggest making it a singleton enum so you don't need to manufacture multiple instances

    public enum VariantComparator implements Comparator<Variant> {
      INSTANCE;
      @Override public int compare(Variant v1, Variant v2) {
        ...
      }
    }